src/perform_step/bdf_perform_step.jl

```diff
@@ -983,3 +983,361 @@ end
+
+    fd_weights = calc_finite_difference_weights(ts_tmp,t+dt,k,Val(max_order))
+    terk = @.. fd_weights[1,k+1] * u
+
+    if u isa Number
+      for i in 2:k+1
+        terk += fd_weights[i,k+1] * u_history[i-1]
+      end
+      terk *= abs(dt^(k))
+    else
+      for i in 2:k+1
+        terk += fd_weights[i,k+1] * u_history[:,i-1]
+      end
+      terk *= abs(dt^(k))
+    end
+
+    atmp = calculate_residuals(_vec(terk), _vec(uprev), _vec(u), integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm, t)
+    cache.terk = integrator.opts.internalnorm(atmp,t)
+    
+    if k > 1
+      fd_weights = calc_finite_difference_weights(ts_tmp,t+dt,k-1,Val(max_order))
+      terkm1 = fd_weights[1,k] * u
+
+      if u isa Number
+        for i in 2:k
+          terkm1 += fd_weights[i,k] * u_history[i-1]
+        end
+        terkm1 *= abs(dt^(k-1))
+      else
+        for i in 2:k
+          terkm1 += fd_weights[i,k] * u_history[:,i-1]
+        end
+        terkm1 *= abs(dt^(k-1))
+      end
+      atmp = calculate_residuals(_vec(terkm1), _vec(uprev), _vec(u), integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm, t)
+      cache.terkm1 = integrator.opts.internalnorm(atmp,t)
+    end
+    if k > 2
+      fd_weights = calc_finite_difference_weights(ts_tmp,t+dt,k-2,Val(max_order))
+      terkm2 = fd_weights[1,k-1] * u
+
+      if u isa Number
+        for i in 2:k-1
+          terkm2 += fd_weights[i,k-1] * u_history[i-1]
+        end
+        terkm2 *= abs(dt^(k-2))
+      else
+        for i in 2:k-1
+          terkm2 += fd_weights[i,k-1] * u_history[:,i-1]
+        end
+        terkm2 *= abs(dt^(k-2))
+      end
+  
+      atmp = calculate_residuals(_vec(terkm2), _vec(uprev), _vec(u), integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm, t)
+      cache.terkm2 = integrator.opts.internalnorm(atmp,t)
+    end
+    if nconsteps > k+1 && k < max_order
+      atmp = calculate_residuals(_vec(terkp1), _vec(uprev), _vec(u), integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm, t)
+      cache.terkp1 = integrator.opts.internalnorm(atmp,t)
+    else
+      cache.terkp1 = zero(cache.terk)
+    end
+  end
+
+  integrator.fsallast = f(u, p, t+dt)
+  integrator.destats.nf += 1
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+  integrator.u = u
+end
+
+function initialize!(integrator, cache::FBDFCache)
+  integrator.kshortsize = 2
+  integrator.fsalfirst = cache.fsalfirst
+  integrator.fsallast = du_alias_or_new(cache.nlsolver, integrator.fsalfirst)
+  resize!(integrator.k, integrator.kshortsize)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+  integrator.f(integrator.fsalfirst, integrator.uprev, integrator.p, integrator.t)
+  integrator.destats.nf += 1
+end
+
+function perform_step!(integrator, cache::FBDFCache{max_order}, repeat_step=false) where max_order
```

Comment:
why is it parameterized on the max order?

Reply:
The internal coefficients calculation is done on `SArray`s, so `max_order` needs to be a compile time const, so that we know the upper bound of the size of `SArray`s.